controllers/helpers.go

```diff
@@ -299,45 +302,134 @@ func GetSubnetUUID(ctx context.Context, client *prismclientv3.Client, peUUID str
+	// name serves as the highest precedence
+	if imageName != nil {
+		resolvedUUID, err := getImageByName(ctx, client, *imageName)
 		if err != nil {
-			if strings.Contains(fmt.Sprint(err), "ENTITY_NOT_FOUND") {
-				return "", fmt.Errorf("failed to find image with UUID %s: %v", *imageUUID, err)
-			}
+			return "", err
 		}
-		foundImageUUID = *imageIntentResponse.Metadata.UUID
-	} else { // else search by name
-		responseImages, err := client.V3.ListAllImage(ctx, "")
+		return resolvedUUID, nil
+	}
+	if imageUUID != nil {
+		resolvedUUID, err := getImageByUUID(ctx, client, *imageUUID)
 		if err != nil {
 			return "", err
 		}
-		// Validate filtered Images
-		foundImages := make([]*prismclientv3.ImageIntentResponse, 0)
-		for _, s := range responseImages.Entities {
-			imageSpec := s.Spec
-			if strings.EqualFold(*imageSpec.Name, *imageName) {
-				foundImages = append(foundImages, s)
-			}
+		return resolvedUUID, nil
+	}
+	if imageLookupFormat != nil && imageLookupBaseOS != nil && k8sVersion != nil {
+		resolvedUUID, err := getImageWithFormat(ctx, client, imageLookupFormat, imageLookupBaseOS, k8sVersion)
+		if err != nil {
+			return "", err
 		}
-		if len(foundImages) == 0 {
-			return "", fmt.Errorf("failed to retrieve image by name %s", *imageName)
-		} else if len(foundImages) > 1 {
-			return "", fmt.Errorf("more than one image found with name %s", *imageName)
-		} else {
-			foundImageUUID = *foundImages[0].Metadata.UUID
+		return resolvedUUID, nil
+	}
+	return "", fmt.Errorf("failed to retrieve image by name or uuid. Verify input parameters")
+}
+
+type ImageLookup struct {
+	BaseOS     string
+	K8sVersion string
+}
+
+func getImageWithFormat(
+	ctx context.Context,
+	client *prismclientv3.Client,
+	imageTemplate,
+	imageLookupBaseOS,
+	k8sVersion *string,
+) (string, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if strings.Contains(*k8sVersion, "v") {
+		k8sVersion = ptr.To(strings.Replace(*k8sVersion, "v", "", 1))
+	}
+	params := ImageLookup{*imageLookupBaseOS, *k8sVersion}
+	t, err := template.New("k8sTemplate").Parse(*imageTemplate)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse template given %s %v", *imageTemplate, err)
+	}
+	var templateBytes bytes.Buffer
+	err = t.Execute(&templateBytes, params)
+	if err != nil {
+		return "", fmt.Errorf(
+			"failed to substitute string %s with params %v error: %w",
+			*imageTemplate,
+			params,
+			err,
+		)
+	}
+	filterString := templateBytes.String()
+	filter := fmt.Sprintf("name==^%s$", filterString)
+	responseImages, err := client.V3.ListAllImage(ctx, filter)
+	if err != nil {
+		return "", fmt.Errorf("failed to list images with error %w", err)
+	}
+	sorted := sortImages(responseImages.Entities)
+	if len(sorted) == 0 {
+		return "", fmt.Errorf("failed to find image with filter %s", filter)
+	}
+	return *sorted[0].Metadata.UUID, nil
+}
+
+func sortImages(images []*prismclientv3.ImageIntentResponse) []*prismclientv3.ImageIntentResponse {
+	sort.Slice(images, func(i, j int) bool {
```

Comment:
can we sort by "created date" and use the latest one?

Reply:
in our case, the name includes the timestamp but we can change this function to explicitly use the date.